core/chains/evm/client/rpc_client.go

```diff
@@ -183,8 +155,82 @@ func NewRPCClient(
 	return r
 }
 
+func (r *RpcClient) SubscribeToHeads(ctx context.Context) (ch <-chan *evmtypes.Head, sub commontypes.Subscription, err error) {
+	ctx, cancel, chStopInFlight, ws, _ := r.acquireQueryCtx(ctx, r.rpcTimeout)
+	defer cancel()
+
+	args := []interface{}{rpcSubscriptionMethodNewHeads}
+	start := time.Now()
+	lggr := r.newRqLggr().With("args", args)
+
+	lggr.Debug("RPC call: evmclient.Client#EthSubscribe")
+	defer func() {
+		duration := time.Since(start)
+		r.logResult(lggr, err, duration, r.getRPCDomain(), "EthSubscribe")
+		err = r.wrapWS(err)
+	}()
+
+	channel := make(chan *evmtypes.Head)
+	forwarder := newSubForwarder(channel, func(head *evmtypes.Head) *evmtypes.Head {
+		head.EVMChainID = ubig.New(r.chainID)
+		r.onNewHead(ctx, chStopInFlight, head)
+		return head
+	}, r.wrapRPCClientError)
+
+	err = forwarder.start(ws.rpc.EthSubscribe(ctx, forwarder.srcCh, args...))
+	if err != nil {
+		return nil, nil, err
+	}
+
+	err = r.registerSub(forwarder, chStopInFlight)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return channel, forwarder, err
+}
+
+func (r *RpcClient) SubscribeToFinalizedHeads(_ context.Context) (<-chan *evmtypes.Head, commontypes.Subscription, error) {
+	interval := r.cfg.FinalizedBlockPollInterval()
+	if interval == 0 {
+		return nil, nil, errors.New("FinalizedBlockPollInterval is 0")
+	}
+	timeout := interval
+	poller, channel := commonclient.NewPoller[*evmtypes.Head](interval, r.LatestFinalizedBlock, timeout, r.rpcLog)
+	if err := poller.Start(); err != nil {
+		return nil, nil, err
+	}
+	return channel, &poller, nil
+}
+
+func (r *RpcClient) Ping(ctx context.Context) error {
+	version, err := r.ClientVersion(ctx)
+	if err != nil {
+		return fmt.Errorf("ping failed: %v", err)
+	}
+	r.rpcLog.Debugf("ping client version: %s", version)
+	return err
+}
+
+func (r *RpcClient) UnsubscribeAllExcept(subs ...commontypes.Subscription) {
+	r.stateMu.Lock()
+	defer r.stateMu.Unlock()
+
+	keepSubs := map[commontypes.Subscription]struct{}{}
+	for _, sub := range subs {
+		keepSubs[sub] = struct{}{}
+	}
+
+	for _, sub := range r.subs {
+		if _, keep := keepSubs[sub]; !keep {
+			sub.Unsubscribe()
+		}
+	}
+	r.latestChainInfo = commonclient.ChainInfo{}
```

Comment:
Hm, there seems to be some misunderstanding.

 `UnsubscribeAllExcept` is expected to be called in two cases: 
-  when we want to disconnect all external callers except aliveLoop (e.g. when RPC is healthy but we just want to change currently active RPC)
- when we want to disconnect all subscriptions

So it does not make sense to reset `latestChainInfo` as we only want to reset it when rpc is unhealthy. 

When Node sees that RPC is unhealthy it should call `rpc.Close` and rpc.Close should contain logic of the `DisconnectAll` method

Reply:
Sounds good, will move this logic to `Close`.